File: 2024/python/p04.py
# Advent of Code 2024 - Day 4

type Coord = tuple[int, int]
type Board = dict[Board, chr]

with open("../input/04.txt") as f:
    raw_data = f.read()

raw_test_data = """MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX"""

def process(s: str) -> Board:
    return {(x,y): c for y, line in enumerate(s.splitlines())
                     for x, c in enumerate(line)}

data = process(raw_data)
test_data = process(raw_test_data)

def compose(*funcs):
    def f(x):
        for f in reversed(funcs):
            x = f(x)
        return x
    return f


def north(loc: Coord) -> Coord:
    x, y = loc
    return (x, y-1)

def west(loc: Coord) -> Coord:
    x, y = loc
    return (x-1, y)

def east(loc: Coord) -> Coord:
    x, y = loc
    return (x+1, y)

def south(loc: Coord) -> Coord:
    x, y = loc
    return (x, y+1)

northwest = compose(north, west)
northeast = compose(north, east)
southwest = compose(south, west)
southeast = compose(south, east)

def match(board, loc, direction, query='XMAS'):
    for letter in query:
        if board.get(loc) != letter:
            return False
        loc = direction(loc)
    return True

def count_xmas(board: Board) -> int:
    count = 0
    for start in board:
        for direction in (north, south, east, west, northwest, northeast, southwest, southeast):
            if match(board, start, direction):
                count += 1
    return count

assert count_xmas(test_data) == 18, "Failed Part 1 test"
ans1 = count_xmas(data)
assert ans1 == 2591, "Failed part 1!"

## Part 2
## find the X-MASes

def match_x_mas(board: Board, start: Coord) -> bool:
    get = lambda fn: board.get(fn(start))
    identity = lambda x: x

    if get(identity) != 'A':
        return False

    if get(northwest) == 'M':
        if get(southeast) != 'S':
            return False
    elif get(southeast) == 'M':
        if get(northwest) != 'S':
            return False
    else:
        return False

    if get(northeast) == 'M':
        if get(southwest) != 'S':
            return False
    elif get(southwest) == 'M':
        if get(northeast) != 'S':
            return False
    else:
        return False

    return True


def count_x_mas(board: Board) -> int:
    count = 0
    for start in board:
        if match_x_mas(board, start):
            count += 1
    return count

assert count_x_mas(test_data) == 9, "Failed part 2 test!"
ans2 = count_x_mas(data)
assert ans2 == 1880, "Failed part 2!"



if __name__ == "__main__":
    print(f"Answer 1: {ans1}")
    print(f"Answer 2: {ans2}")

File: 2024/clojure/p04.clj
;; # 🎄 Advent of Code - 2024 - Day 4

(ns p04
  [:require [clojure.test :as test]
            [clojure.string :as str]])

;; ## Data
;; The data is a field of letters, we'll turn this into a map.

(def data-string (slurp "../input/04.txt"))
(def test-data-string "MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX")

(defn process [s]
  (into {}
    (for [[y line] (map-indexed vector (str/split-lines s))
          [x c] (map-indexed vector line)]
       [[x y] c])))

(def data (process data-string))
(def test-data (process test-data-string))

;; ## Logic
;; To solve this, we'll create the directions
;; and write a match utility that checks to see
;; if we have an X-MAS at each location and direction.

(defn north [[x y]] [x (dec y)])
(defn south [[x y]] [x (inc y)])
(defn east [[x y]] [(inc x) y])
(defn west [[x y]] [(dec x) y])
(def north-west (comp north west))
(def north-east (comp north east))
(def south-west (comp south west))
(def south-east (comp south east))

(def directions 
  [north south east west 
   north-west north-east 
   south-west south-east])

(defn match 
  ([m loc dir] (every? true? (map = "XMAS" (map m (iterate dir loc)))))
  ([m loc] (count (filter (partial match m loc) directions)))
  ([m] (transduce
         (map (partial match m)) 
         +
         (keys m))))

(def part-1 match)
(def ans1 (part-1 data))

(test/deftest test-part-1
  (test/is (= 18 (part-1 test-data)))
  (test/is (= 2591 ans1)))

;; ## Part 2
;; Instead of finding XMAS in a row, now we are supposed to look
;; for crosses of MAS.

(defn match-x-mas 
  ([m loc] (and
             (= (m loc) \A)
             (or (and (= (m (north-east loc)) \M)
                      (= (m (south-west loc)) \S))
                 (and (= (m (south-west loc)) \M)
                      (= (m (north-east loc)) \S)))
             (or (and (= (m (north-west loc)) \M)
                      (= (m (south-east loc)) \S))
                 (and (= (m (south-east loc)) \M)
                      (= (m (north-west loc)) \S)))))
  ([m] (count (filter (partial match-x-mas m) (keys m)))))

(def part-2 match-x-mas)
(def ans2 (part-2 data))

(test/deftest test-part-2
  (test/is (= 9 (part-2 test-data)))
  (test/is (= 1880 ans2)))

;; ## Main

(comment
  (test/run-tests))

(defn -main [& args]
  (println "Answer1:" ans1)
  (println "Answer2:" ans2))
